test/adapters/level_zero/event_cache_tests.cpp

```diff
@@ -0,0 +1,170 @@
+        for (int j = 0; j < numEnqueues; j++) {
+            enqueueWork(nullptr, i * numEnqueues + j);
+        }
+        ASSERT_SUCCESS_OR_UNSUPPORTED(urQueueFinish(queue));
+        verifyData();
+    }
+
+    // TODO: why events are not reused for UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE?
+    if ((flags & UR_QUEUE_FLAG_DISCARD_EVENTS) &&
+        !(flags & UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE)) {
+        ASSERT_EQ((*ZeCallCount)["zeEventCreate"], 2);
+    } else {
+        ASSERT_GE((*ZeCallCount)["zeEventCreate"], numIters * numEnqueues);
+    }
+}
+
+TEST_P(urEventCacheTest, eventsReuseWithVisibleEvent) {
+    static constexpr int numIters = 16;
+    static constexpr int numEnqueues = 128;
+
+    for (int i = 0; i < numIters; i++) {
+        std::vector<uur::raii::Event> events(numEnqueues);
+        for (int j = 0; j < numEnqueues; j++) {
+            enqueueWork(events[j].ptr(), i * numEnqueues + j);
+        }
+        ASSERT_SUCCESS_OR_UNSUPPORTED(urQueueFinish(queue));
+        verifyData();
+    }
+
+    ASSERT_LT((*ZeCallCount)["zeEventCreate"], numIters * numEnqueues);
+}
+
+TEST_P(urEventCacheTest, eventsReuseWithVisibleEventAndWait) {
+    static constexpr int numIters = 16;
+    static constexpr int numEnqueues = 128;
+    static constexpr int waitEveryN = 16;
+
+    for (int i = 0; i < numIters; i++) {
+        std::vector<uur::raii::Event> events;
+        for (int j = 0; j < numEnqueues; j++) {
+            events.emplace_back();
+            enqueueWork(events.back().ptr(), i * numEnqueues + j);
+
+            if (j > 0 && j % waitEveryN == 0) {
+                ASSERT_SUCCESS(urEventWait(waitEveryN,
+                                           (ur_event_handle_t *)events.data()));
+                verifyData();
+                events.clear();
+            }
+        }
+        ASSERT_SUCCESS_OR_UNSUPPORTED(urQueueFinish(queue));
+    }
+
+    ASSERT_GE((*ZeCallCount)["zeEventCreate"], waitEveryN);
+    // TODO: why there are more events than this?
+    // ASSERT_LE((*ZeCallCount)["zeEventCreate"],  waitEveryN * 2 + 2);
+}
+
+template <typename T>
+inline std::string
+printFlags(const testing::TestParamInfo<typename T::ParamType> &info) {
+    const auto device_handle = std::get<0>(info.param);
+    const auto platform_device_name =
+        uur::GetPlatformAndDeviceName(device_handle);
+    auto flags = combineFlags(std::get<1>(info.param));
+
+    std::stringstream ss;
+    ur::details::printFlag<ur_queue_flag_t>(ss, flags);
+
+    auto str = ss.str();
+    std::replace(str.begin(), str.end(), ' ', '_');
+    std::replace(str.begin(), str.end(), '|', '_');
+    return platform_device_name + "__" + str;
+}
+
+UUR_TEST_SUITE_P(
+    urEventCacheTest,
+    ::testing::Combine(
+        testing::Values(0, UR_QUEUE_FLAG_DISCARD_EVENTS),
+        testing::Values(0, UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE),
+        // TODO: why the test fails with UR_QUEUE_FLAG_SUBMISSION_BATCHED?
```

Comment:
with what combination of the flags? The discard events path doesn't seem that well tested, so it's not inconceivable that's its broken when a few different flags are set.

Reply:
I rerun this on the latest driver and everything works fine there so I guess we can just uncomment this once we switch our CI to the newer driver.